File: vistools/src/main/java/mapvis/vistools/colormap/ColorBar.java
package mapvis.vistools.colormap;

import javax.swing.*;
import java.awt.*;
import java.util.function.Function;

public class ColorBar extends JComponent {
    private Function<Double, Color> colorFunc;

    public ColorBar(Function<Double, Color> colorFunc) {
        this.colorFunc = colorFunc;
        setPreferredSize(new Dimension(100, 20));
    }

    @Override
    public void paintComponents(Graphics g) {
        super.paintComponents(g);
        if (g instanceof Graphics2D)
            return;

        g.setColor(getBackground());
        g.fillRect(0, 0, getWidth(), getHeight());

        double slope = getWidth() / 256.0;

        int n = 255;
        while(n-- != 0) {
            int x = (int)(n * slope);
            g.setColor(colorFunc.apply(n / 255.0));
            g.fillRect(x, 0, (int)slope, getHeight());
        }
    }
}

File: vistools/src/main/kotlin/mapvis/vistools/colormap/ColorBar.kt
package mapvis.vistools.colormap

import javax.swing.*
import java.awt.*
import java.util.function.Function

public class Colorbar(val colorFunc: Function<Double, Color>) : JComponent() {
    {
        this.setPreferredSize(Dimension(100, 20))
    }

    override fun paintComponent(g: Graphics) {
        if (g !is Graphics2D)
            return;

        g.setColor(getBackground())
        g.fillRect(0, 0, getWidth(), getHeight())

        val slope = getWidth().toFloat() / 256.0.toFloat()

        for (n in 0..255) {
            val x = (n.toFloat() * slope).toInt()
            g.setColor(colorFunc.apply(n.toDouble() / 255.0))
            g.fillRect(x, 0, slope.toInt(), getHeight())
        }
    }
}